flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java

```diff
@@ -804,16 +850,31 @@ private void fulfillPendingSlotRequestWithPendingTaskManagerSlot(PendingSlotRequ
 		return Optional.empty();
 	}
 
-	private boolean isFulfillableByRegisteredSlots(ResourceProfile resourceProfile) {
+	private boolean isFulfillableByRegisteredOrPendingSlots(ResourceProfile resourceProfile) {
 		for (TaskManagerSlot slot : slots.values()) {
 			if (slot.getResourceProfile().isMatching(resourceProfile)) {
 				return true;
 			}
 		}
+
+		for (PendingTaskManagerSlot slot : pendingSlots.values()) {
```

Comment:
I am not very familiar with the `SlotManager`. Can you explain me why we did not do this check/loop before?

Reply:
Before this change:
- For Yarn/K8s mode, as long as the ResourceProfile of the `SlotRequest` could fit in the `defaultSlotResourceProfile`, the SlotManager would always trigger `allocateResource` and get `pendingSlot`.
- For Standalone mode, it will set `failUnfulfillableRequest` to false for a startup period to wait there is registered slot in the cluster. Since the `ResourceManager#allocateResource` would always return false in standalone mode, this method checks whether the ResourceProfile of the `SlotRequest` could fit in.

After this change:
- For Yarn/K8s mode, even if the ResourceProfile of the `SlotRequest` could fit in the `defaultSlotResourceProfile`, the `allocateResource` could return no `pendingSlot` when the total number of slot reaches max limit. In this case, the "resource fit in" check should be move to `isFulfillableByRegisteredOrPendingSlots` function.
- For Standalone mode, since there is no `pendingSlot`, we do not break the origin behavior.

Reply:
To summarize:
- Before this change, this method do the "resource fit" check mainly for Standalone mode. There is no `pendingSlot` in this mode.
- After this change, this method could also do the "resource fit" check for Yarn/K8s when the number of slots reached max limit. The `pendingSlot` may not register back quickly at the startup period. Only check the registered `slots` would cause a false alarm.

Reply:
Thanks for the explanation.

Reply:
Which test covers this change?

Reply:
Actually, `testMaxSlotLimitAllocateResource` already cover it. In its logic, we do not register any TM to the SM, so the `slots` should be empty. When the second `SlotReport` comes in, the `allocateResource` would return false, and the logic will go into this check.

Reply:
alright, makes sense